digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/process/strategy/TrackCalculatorStrategy.scala

```diff
@@ -205,28 +206,58 @@ trait TrackCalculatorStrategy {
       }
   }
 
-  protected def setCalibrationPoint(pl: ProjectLink, userCalibrationPoint: Option[UserDefinedCalibrationPoint], startCP: Boolean, endCP: Boolean, source: CalibrationPointSource = UnknownSource) = {
+  protected def setCalibrationPoint(pl: ProjectLink, userCalibrationPoint: Option[UserDefinedCalibrationPoint],
+                                    startCP: Boolean, endCP: Boolean, source: CalibrationPointSource = UnknownSource): ProjectLink = {
     val sCP = if (startCP) CalibrationPointsUtils.makeStartCP(pl) else None
     val eCP = if (endCP) CalibrationPointsUtils.makeEndCP(pl, userCalibrationPoint) else None
     pl.copy(calibrationPoints = CalibrationPointsUtils.toProjectLinkCalibrationPointsWithSourceInfo((sCP, eCP), source))
   }
 
-  protected def getUntilNearestAddress(seq: Seq[ProjectLink], endProjectLink: ProjectLink): (Seq[ProjectLink], Seq[ProjectLink]) = {
-    if (endProjectLink.discontinuity == MinorDiscontinuity || endProjectLink.discontinuity == Discontinuous) {
-      val continuousProjectLinks = seq.takeWhile(pl => pl.startAddrMValue < endProjectLink.endAddrMValue)
+  private def lastSegmentDirection(pl: ProjectLink): Vector3d = {
+    if ((pl.sideCode == SideCode.TowardsDigitizing && pl.reversed == false) ||
+      (pl.sideCode == SideCode.AgainstDigitizing && pl.reversed == true)) {
+      GeometryUtils.lastSegmentDirection(pl.geometry)
+    } else {
+      GeometryUtils.firstSegmentDirection(pl.geometry)
+    }
+  }
 
-      if (continuousProjectLinks.isEmpty)
+  /**
+    * Returns project links for the other track before and after the point where there is discontinuity on the track.
+    *
+    * @param trackA
+    * @param linkOnTrackB
+    * @return (Project links before the discontinuity point, project links after the discontinuity point)
+    */
+  protected def getUntilNearestAddress(trackA: Seq[ProjectLink], linkOnTrackB: ProjectLink): (Seq[ProjectLink], Seq[ProjectLink]) = {
+    if (linkOnTrackB.discontinuity == MinorDiscontinuity || linkOnTrackB.discontinuity == Discontinuous) {
+
+      // Sort links in order by the distance from the end of the link on track b.
+      val linkOnTrackBEndPoint = linkOnTrackB.getEndPoints()._2
+      val nearestLinks: Seq[(ProjectLink, Double)] = trackA.map(pl => (pl,
+        pl.getEndPoints()._2.distance2DTo(linkOnTrackBEndPoint)
+      )).sortWith(_._2 < _._2) // Links with nearest end points first
+
+      // Choose the nearest link that has similar enough angle to the linkOnTrackB and set it as lastLink
+      val linkOnTrackBDirection = lastSegmentDirection(linkOnTrackB)
+      val maxAngleBetweenLinks = math.toRadians(45.0)
```

Comment:
This 45 degrees might not be a correct value here. I'll continue testing and finding better value here.